File: Mathematical Algorithms/Succinct Encoding of Binary Tree/SEBT.py
class Node:
    # Utility function to create new Node
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None


def EncodeSuccinct(root, struc, data):
    # If root is None , put 0 in structure array and return
    if root is None:
        struc.append(0)
        return

    # Else place 1 in structure array, key in 'data' array
    # and recur for left and right children
    struc.append(1)
    data.append(root.key)
    EncodeSuccinct(root.left, struc, data)
    EncodeSuccinct(root.right, struc, data)


def DecodeSuccinct(struc, data):
    if len(struc) <= 0:
        return None

    # Remove one item from structure list
    b = struc[0]
    struc.pop(0)

    # If removed bit is 1
    if b == 1:
        key = data[0]
        data.pop(0)

        # Create a tree node with removed data
        root = Node(key)

        # And recur to create left and right subtrees
        root.left = DecodeSuccinct(struc, data)
        root.right = DecodeSuccinct(struc, data)
        return root

    return None


def preorder(root):
    if root is not None:
        print("key: %d" % root.key, end=" ")

        if root.left is not None:
            print("| left child: %d" % root.left.key, end=" ")
        if root.right is not None:
            print("| right child %d" % root.right.key, end=" ")
        print()
        preorder(root.left)
        preorder(root.right)

File: Mathematical Algorithms/Succinct Encoding of Binary Tree/SEBT.java
import java.util.Vector;

public class SEBT {
    static class Node {
        int key;
        Node left, right;
    }

    static Node newNode(int key) {
        Node temp = new Node();
        temp.key = key;
        temp.left = temp.right = null;
        return (temp);
    }

    static Vector<Boolean> struc;
    static Vector<Integer> data;
    static Node root;

    static void EncodeSuccinct(Node root) {

        // If root is null, put 0 in
        // structure array and return
        if (root == null) {
            struc.add(false);
            return;
        }

        // Else place 1 in structure array,
        // key in 'data' array and recur
        // for left and right children
        struc.add(true);
        data.add(root.key);
        EncodeSuccinct(root.left);
        EncodeSuccinct(root.right);
    }

    static Node DecodeSuccinct() {
        if (struc.size() <= 0)
            return null;

        // Remove one item from structure list
        boolean b = struc.get(0);
        struc.remove(0);

        // If removed bit is 1,
        if (b == true) {

            // Remove an item from data list
            int key = data.get(0);
            data.remove(0);

            // Create a tree node with the
            // removed data
            Node root = newNode(key);

            // And recur to create left and
            // right subtrees
            root.left = DecodeSuccinct();
            root.right = DecodeSuccinct();
            return root;
        }
        return null;
    }

    static void preorder(Node root) {
        if (root != null) {
            System.out.print("key: " + root.key);
            if (root.left != null)
                System.out.print(" | left child: " +
                        root.left.key);
            if (root.right != null)
                System.out.print(" | right child: " +
                        root.right.key);
            System.out.println();

            preorder(root.left);
            preorder(root.right);
        }
    }
}
